### ISSUE TITLE
terraform show doesn't properly output child_modules in certain circumstances - with ready fix

### ISSUE THREAD

**Issue by cube2222** [hashicorp/terraform#27351]
Hey, I'll try to keep it short.

Let's say you have 3 modules, in the examples that's the `root` module, the `resourceless` module, and the `lost` module.
The `resourceless` module is a child of the `root` module.
The `lost` module is a child of the `resourceless` module.

The caveat though is that the `resourceless` module doesn't create any resources itself, it only creates the `lost` module.

This will break the recurrence in `jsonstate/state.go` and will never print the contents of the `resourceless` and `lost` modules.

I have a simple fix ready, including a regression test, so will create a PR shortly.

### Terraform Version
All 0.14.x versions and head of master, the problematic code hasn't been touched for a while though, so it should be reproducible in earlier versions too.

### Terraform Configuration Files
```terraform
# main.tf
module "resourceless-module" {
  source = "./resourceless"
}
```

```terraform
# resourceless/main.tf
module "lost-module" {
  source = "./lost"
}
```

```terraform
# resourceless/lost/main.tf
resource random_id "lost-id" {
  byte_length = 16
}
```

### Expected Behavior
Running `terraform show -json` should properly print all the modules in a `child_modules` hierarchy. Like this (this is the output with my fix):
```
{
  "format_version": "0.1",
  "terraform_version": "0.15.0",
  "values": {
    "root_module": {
      "child_modules": [
        {
          "address": "module.resourceless-module",
          "child_modules": [
            {
              "address": "module.resourceless-module.module.lost-module",
              "resources": [
                {
                  "address": "module.resourceless-module.module.lost-module.random_id.lost-id",
                  "mode": "managed",
                  "name": "lost-id",
                  "provider_name": "registry.terraform.io/hashicorp/random",
                  "schema_version": 0,
                  "type": "random_id",
                  "values": {
                    "b64_std": "DI0iBocGJYcaeDpQrDYnlg==",
                    "b64_url": "DI0iBocGJYcaeDpQrDYnlg",
                    "byte_length": 16,
                    "dec": "16683539925555807513425851214500079510",
                    "hex": "0c8d2206870625871a783a50ac362796",
                    "id": "DI0iBocGJYcaeDpQrDYnlg",
                    "keepers": null,
                    "prefix": null
                  }
                }
              ]
            }
          ]
        }
      ]
    }
  }
}
```

### Actual Behavior
```
{"format_version":"0.1","terraform_version":"0.15.0","values":{"root_module":{}}}
```

### Steps to Reproduce
Just apply the above terraform files and run `terraform show -json`

**Comment by danieldreier** [comment]
@cube2222 thank you for the report and clear reproduction case! I've reproduced this on 0.14.2 and 0.13.5. Most of the team of off for the holidays, and this looks real but non-urgent to me, so be aware that it probably won't get much more attention until January. Please let me know if I've misunderstood the issue and and it's causing an urgent incident for you.

**Comment by cube2222** [comment]
Hey, since the Christmas holiday season is over I suppose, could you please take a look at this?
Here's the Pull Request aiming to fix this: #27352

**Comment by cube2222** [comment]
PR merged, thanks!
